### DISCUSSION TITLE
VImage save to svs format very slow on windows, is there any way to accelerate it

### DISCUSSION THREAD

**Discussion by zhusihan-python** [libvips/libvips#3219] [Q&A]
im trying to trans a ITK image to VImage, and save it as a svs format file, as the itk did not support ome tiff or svs directly.

it works, yet writing to disk takes too long, a 20202 * 12133 * 3 size image cost 17 mins.

got warnings like this, does it related to the slow processing?
`(process:40236): VIPS-WARNING **: 10:07:25.970: mozjpeg library likely detected. Disable emission of Huffman tables in JpegTables tag, and use optimize_coding to avoid potential issues`

my environment: windows 10, vip 8.13.2, visual studio 2019, msvc 14, ISO c++ 17

my write function as below
```
static void write_page(const VImage &in, const unsigned tile_size,
                       const std::optional<int> jpeg_quality,
                       PageType page_type, TIFF *out) {
  const uint32_t width = in.width();
  const uint32_t height = in.height();

#ifdef WITH_SPINNER
  spinner->SetText("Generating layer with size (" + std::to_string(width) + ", " + std::to_string(height) + ")");
#endif

  const int tile_width = (page_type == PageType::kStriped) ? width : tile_size;

  VImage cached = in.tilecache(
      VImage::option()
      ->set("tile_width", tile_width)
      ->set("tile_height", static_cast<int>(tile_size))
      ->set("persistent", false));

  init_tiff_page(out, width, height, 0, 0);

  // Set jpeg compression
  TIFFSetField(out, TIFFTAG_COMPRESSION, COMPRESSION_JPEG);
  if (jpeg_quality)
    TIFFSetField(out, TIFFTAG_JPEGQUALITY, *jpeg_quality);

  if (page_type == PageType::kTiled) {
    TIFFSetField(out, TIFFTAG_TILEWIDTH, tile_size);
    TIFFSetField(out, TIFFTAG_TILELENGTH, tile_size);
  } else
    TIFFSetField(out, TIFFTAG_ROWSPERSTRIP, tile_size);

  VipsImageTileGenerator tiles(cached, tile_width, tile_size);
  for (const std::optional<Tile> &tile : tiles) {
    assert(tile);
    const Buffer &buffer = (*tile).buffer;
    const unsigned index = (*tile).index;
    if (page_type == PageType::kTiled)
      TIFFWriteEncodedTile(out, index, buffer.data.get(), buffer.size);
    else
      TIFFWriteEncodedStrip(out, index, buffer.data.get(), buffer.size);
  }
  TIFFWriteDirectory(out);
}
```
new VImage from itk image
```
itk::Image<PixelType, Dimension>::SizeType size = resampleF->GetOutput()->GetLargestPossibleRegion().GetSize();
const void *           dataPtr = resampleF->GetOutput()->GetBufferPointer();
long long unsigned int buffer_len = size[0] * size[1] * 3;
VImage                 in = VImage::new_from_buffer(dataPtr, buffer_len, "svsdata");
```

**Comment by zhusihan-python** [comment]
another choice is to call VImage.tiffsave to write a tiff file, its very fast, but the size of the output  file is too large, suppose a 4 level svs file is 220MB, 10 level tiff file with same contents can reach 900MB,  even1 level tiff file more than 700MB, set VOption compression seems not reduce the file size.

**Reply by jcupitt** [reply]
It should work. With this test image:

https://openslide.cs.cmu.edu/download/openslide-testdata/Aperio/

I see:

```
$ ls -l CMU-1.svs
-rw-r--r-- 1 john john 177552579 Feb 13  2021 CMU-1.svs
$ vipsheader CMU-1.svs
CMU-1.svs: 46000x32914 uchar, 4 bands, srgb, openslideload
$ time vips copy CMU-1.svs x.tiff[compression=jpeg,tile,pyramid]

real	0m10.962s
user	1m13.363s
sys	0m3.843s
$ ls -l x.tiff
-rw-r--r-- 1 john john 137361146 Dec  7 08:54 x.tiff
```

About 10s for the transform, and the file size is smaller. That's with the default Q=75 of course, you'd probably want to raise that.

**Reply by zhusihan-python** [reply]
`vips copy CMU-1.svs x.tiff[compression=jpeg,tile,pyramid]`
this command line works on my svs file, i get a 22MB tiff file from a 220MB svs file.
but when i do it in cpp code like below:
```
VImage in = VImage::new_from_memory(dataPtr, buffer_len, size[0], size[1], 3, VIPS_FORMAT_UCHAR);

const uint32_t width = in.width();
const uint32_t height = in.height();

in.tiffsave(outFilename.c_str(),
             VImage::option()
               ->set("tile_width", TILE_SIZE)
               ->set("tile_height", static_cast<int>(TILE_SIZE))
               ->set("pyramid", true)
               ->set("tile", true)
               ->set("bigtiff", true)
               //->set("compression", COMPRESSION_JPEG)
               ->set("Q", 70));
```
if set compression, i only get an empty tiff file, if remove compression, i got a 900M tiff file, if the full code is needed, i will pick them to a single file and upload it here later
im using libvips vips-dev-w64-all-8.13.2.zip

**Reply by jcupitt** [reply]
`COMPRESSION_JPEG` is the libtiff `#define`, you need to use the libvips enum, `VIPS_FOREIGN_TIFF_COMPRESSION_JPEG`.

I usually use strings, they are safer than an untyped int. I would write:

```C++
/* compile with
 *
 *      g++ -g -Wall try336.cpp `pkg-config vips-cpp --cflags --libs`
 */

#include <vips/vips8>

using namespace vips;

int
main( int argc, char **argv )
{     
        if (VIPS_INIT (argv[0]))
                vips_error_exit (NULL);

        VImage image = VImage::new_from_file(argv[1], VImage::option()
                ->set("access", "sequential"));
  
        image.write_to_file(argv[2], VImage::option()
                ->set("tile-width", 256)
                ->set("tile-height", 256)
                ->set("tile", true)
                ->set("pyramid", true)
                ->set("bigtiff", true)
                ->set("compression", "jpeg")
                ->set("Q", 80));

        return 0;
}
```

I see:

```
$ g++ -g -Wall try336.cpp `pkg-config vips-cpp --cflags --libs`
$ time ./a.out ~/pics/openslide/CMU-1.svs x.tiff

real	0m10.288s
user	1m9.860s
sys	0m3.110s
$ ls -l x.tiff
-rw-r--r-- 1 john john 150537862 Dec  7 09:59 x.tiff
```

**Reply by zhusihan-python** [reply]
thanks, it finally worked. the quality 80 get a really small tiff.
i think 4 level is enough in my case, does downsample level support being specified, it get a 8 or 9 levels in fact which will cost extra writing time, the level or depth param can not achieve that as I tested.

**Reply by jcupitt** [reply]
I'd use 85, I think.

You can write a smaller pyramid by just calling resize before write_to_file(), eg.:

```C++
image
    .resize(0.25)
    .write_to_file("x.tif", ...);
```

Will skip the largest two levels.

**Comment by jcupitt** [comment]
Hello @zhusihan-python,

libvips uses openslide to load SVS and openslide is read-only, there is no support for SVS write. 

As you say, your best bet would probably be to write as tiff. Tiled jpeg-compressed pyramidal tiff is very similar to svs, and should be supported by most systems. libvips also supports jpeg2000 write, so that might also help.

You'd need to share many more details about your system to diagnose the speed problem:

-  libvips is demand-driven, so low speed can be caused by any part of the libvips pipeline. You'd need to provide a complete, runnable program I could test.
- Which libvips windows binary did you download? You need the `-all` version for openslide support.
- How did you build the C++ binding?
- I don't really understand your write function. Are you generating the svs yourself and running a libvips pipeline for every tile? That will be extremely slow.

**Comment by zhusihan-python** [comment]
i build the C++ binding using visual studio 2019, msvc 14
i do want to generate a svs from a image array by myself, i did that by tifffile in python which turns out not bad performance, but tifffile is pure python does not  provide a cpp interface

```
def gen_data(source, row, col, h, w):
    while True:
        for i in range(row):
            for j in range(col):
                im = source[i * h:(i + 1) * h, j * w:(j + 1) * w, :]
                yield im

# image_array is a numpy array
image_height = image_array.shape[0]
image_width = image_array.shape[1]
row = math.floor(image_height/HEIGHT)
col = math.floor(image_width/WIDTH)

mpp = 0.25
mag = 20
filename = "out.svs"

height_cut = image_height - image_height % HEIGHT
width_cut = image_width - image_width % WIDTH
multi_hw = np.int64([(height_cut, width_cut), (height_cut/4, width_cut/4), (height_cut/16, width_cut/16)])
thw = [(1024, 1024), (256, 256), (64, 64)]
steps = [1, 4, 16]

svs_desc = f"PSDS 1000 Image\nABC |AppMag = {mag}|Filename = {filename}|MPP = {mpp}"
with tifffile.TiffWriter(out_file, bigtiff=True) as tif:

    compression = ['JPEG', 95, dict(outcolorspace='YCbCr')]
    kwargs = dict(subifds=None, photometric='rgb', planarconfig='CONTIG', compression=compression,
                  dtype=np.uint8,
                  metadata=None)
    for i, hw in enumerate(multi_hw):
        step = steps[i]
        if step > 1:
            source = cv2.resize(image_array, (image_array.shape[1]//steps[i], image_array.shape[0]//steps[i]))
        else:
            source = image_array
        gen = gen_data(source, row, col, thw[i][0], thw[i][1])

        if i == 0:
            desc = svs_desc.format(mag=20, filename='test', mpp=0.35)
            tif.write(data=gen, shape=(*hw, 3), tile=thw[i],
                      description=desc, subfiletype=0, **kwargs)
            thumbnail_im = cv2.resize(image_array, (image_array.shape[1]//20, image_array.shape[0]//20))
            tif.write(data=thumbnail_im, description='', **kwargs)
        else:
            tif.write(data=gen, shape=(*hw, 3), tile=thw[i],
                      description='', subfiletype=1, **kwargs)
```
the difference i can tell is that in python code i write 3 times equals the level counts, in cpp code i write thousands of times equals the tile counts, but i have no idea how do write a whole level each time in cpp, as i did not understands the data structures in the deep end actually
